client/asset/eth/eth.go

```diff
@@ -313,28 +318,151 @@ func (*ExchangeWallet) PreRedeem(req *asset.PreRedeemForm) (*asset.PreRedeem, er
 	return nil, asset.ErrNotImplemented
 }
 
+// EthCoin implements the asset.Coin interface for ETH
+type EthCoin struct {
+	address common.Address
+	value   uint64
+}
+
+// address + value(uint64)
+const ethCoinIdLength = common.AddressLength + 8
+
+func (c EthCoin) ID() dex.Bytes {
+	id := make([]byte, ethCoinIdLength)
+	copy(id[:common.AddressLength], c.address.Bytes())
+	binary.LittleEndian.PutUint64(id[common.AddressLength:], c.value)
+	return id
+}
+
+func (c EthCoin) String() string {
+	return fmt.Sprintf("%v: %v", c.address.String(), c.value)
+}
+
+func (c EthCoin) Value() uint64 {
+	return c.value
+}
+
+func ethCoinFromID(id dex.Bytes) (*EthCoin, error) {
+	if len(id) != ethCoinIdLength {
+		return nil, fmt.Errorf("coin id should be 28 bytes but got %v",
+			len(id))
+	}
+
+	var address common.Address
+	copy(address[:], id[:common.AddressLength])
+
+	return &EthCoin{
+		address: address,
+		value:   binary.LittleEndian.Uint64(id[common.AddressLength:]),
+	}, nil
+}
+
 // FundOrder selects coins for use in an order. The coins will be locked, and
 // will not be returned in subsequent calls to FundOrder or calculated in calls
 // to Available, unless they are unlocked with ReturnCoins.
-// The returned []dex.Bytes contains the redeem scripts for the selected coins.
-// Equal number of coins and redeemed scripts must be returned. A nil or empty
-// dex.Bytes should be appended to the redeem scripts collection for coins with
-// no redeem script.
-func (*ExchangeWallet) FundOrder(ord *asset.Order) (asset.Coins, []dex.Bytes, error) {
-	return nil, nil, asset.ErrNotImplemented
+// In UTXO based coins, the returned []dex.Bytes contains the redeem scripts for the
+// selected coins, but since there are no redeem scripts in Ethereum, the coin ids
+// are returned there.
+// Equal number of coins and redeem scripts must be returned.
+func (eth *ExchangeWallet) FundOrder(ord *asset.Order) (asset.Coins, []dex.Bytes, error) {
+	maxFees := ord.DEXConfig.MaxFeeRate * ord.DEXConfig.SwapSize * ord.MaxSwapCount
+	fundsNeeded := ord.Value + maxFees
+	balance, err := eth.Balance()
+	if err != nil {
+		return nil, nil, err
+	}
+
+	if balance.Available < fundsNeeded {
+		return nil, nil, fmt.Errorf("%v gwei are needed for order, but address %v only has %d available",
+			fundsNeeded, eth.acct.Address.Hex(), balance.Available)
```

Comment:
Sorry to keep jumping in on this review, there are some gotchas with `%x` that comes up a lot.  One is that if the type in the arg list is also a `Stringer`, which `common.Address` is, it's going to first evaluate the `String` method, then print the hexadecimal representation of *of the bytes of the string* not the hexadecimal representation of the bytes.

With `common.Address` I don't know if this will also happen, maybe the `Hex` method it provides helps, but `%x` will change into a non-standard formatting. i.e.

```
  b6de8bb5ed28e6be6d671975cad20c03931be981
vs
0xB6De8BB5ed28E6bE6d671975cad20C03931bE981
```

Just consider usage of `%x` carefully. It has bit us before.  https://play.golang.org/p/5AO3hmMRReH

Reply:
Oh, interesting I discovered that `common.Address` is a `fmt.Formatter` so the printf code has yet another special path to call `(Address).Format` and do a basic (and different) hex conversion itself rather than letting the fmt package just call it's Stringer and mess it up.  Fun. Anyway, it is a different formatting so it's still something to consider.

Reply:
It looks good with the `Format` function.

Reply:
I don't know enough about this to say if it's an issue, but to compliant with [EIP-55 Mixed-case checksum address encoding](https://eips.ethereum.org/EIPS/eip-55), it would have to use `Hex` or `String` rather than this fallback Formatter.
You guys have a feel for the importance of this?

Reply:
The format function uses the checksum hex:
https://github.com/ethereum/go-ethereum/blob/b522f5e091fe5e09056ac96c890f9ec5aac5d88d/common/types.go#L277

What I have now is what you have on line 29 of your playground file. I think it looks right.

Reply:
No joke, formatting in go is hard. So no %x then, I may be doing that somewhere.

Reply:
This took me by surprise. The behavior of a slice is different, and I would have expected `[:]` to give me what the other types do. https://play.golang.org/p/kccYaGSyCGL